### DISCUSSION TITLE
CircuitPython

### DISCUSSION THREAD

**Discussion by tannewt** [aappleby/hancho#3] [General]
Thanks for Hancho! I've been looking for a new build system for CircuitPython and think this could be it. I like that it's all simple Python! I've started converting our make files to hancho here: https://github.com/tannewt/circuitpython/tree/clang18 (we have a discord too if you want to chat: https://adafru.it/discord)

I'm finding that our build system is pretty inside out. We build thousands of binaries for different microcontroller boards. We have high level rules we share between all of circuitpython and then narrower rules per microcontroller family and finally specific customization per board.

I've started setting it up by having the "top level" hancho file in the board directory and then loading the outer hancho files from there. Then it'll have one top level build directory that will have unique artifacts per board and shared object files. The main challenge I have so far is that the functions are run from within the deepest directory. So, it only works if I use absolute paths for everything.

Is this something you envision supporting with Hancho or should I modify it to work for us? Any tips would be appreciated.

**Comment by aappleby** [comment]
Hi, I'll need to look at your build setup to understand what you mean by "from within the deepest directory" exactly.

I'm not really hard-set on any particular behavior regarding subdirectories and subprojects, aside from "glob() should search the directory of the hancho file that called it".

**Reply by tannewt** [reply]
I've got [three files currently](https://github.com/tannewt/circuitpython/commit/72ee3697c195772c66065f47b6c2d0002fddaddb):
* circuitpython.hancho
* ports/rp2040/rp2040.hancho
* ports/rp2040/boards/adafruit_feather_rp2040/board.hancho

I'm running `python tools/hancho/hancho.py ports/rp2040/boards/adafruit_feather_rp2040/board.hancho`. board.hancho then calls a function from rp2040.hancho and it a function from circuitpython.hancho.

I agree with that glob idea but am running into the problem that the functions are run from the `board.hancho` context. Is it too weird to wrap any functions from a module with chdir as well?

Any rules called from the functions also have the wrong current directory. For example circuitpython.hancho has a Rule that runs a python script so the `script = Rule()` is at the top level of the module but the `script()` is in the function which gives it the wrong working directory.

**Comment by aappleby** [comment]
OK, I'll try and repro in a bit.

I would rather not automagically wrap functions with chdir() stuff, but I suppose it could be done in the Hancho module loader if it has to.

**Comment by aappleby** [comment]
aappleby@lappytron:~/circuitpython$ make fetch-all-submodules
python3 tools/ci_fetch_deps.py all
::group::Sync submodule URLs
git submodule sync --quiet (in /home/aappleby/circuitpython)
::endgroup::
Duration: 0.5187677610001629
Target: all
Submodules: . frozen
::group::Init submodules (using depth)
git submodule update --init --depth 1 . frozen (in /home/aappleby/circuitpython)
Cloning into '/home/aappleby/circuitpython/tools/hancho'...
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

???

**Comment by aappleby** [comment]
Oh nm it's just the .gitmodules file

**Comment by aappleby** [comment]
I'm trying to follow BUILDING.md and hitting a few bumps in your fork - I need to put this aside for a bit but I'll try and look into it more later. 

aappleby@lappytron:~/circuitpython$ make -C mpy-cross
make: Entering directory '/home/aappleby/circuitpython/mpy-cross'
- Verbosity options: any combination of "steps commands rules", as `make V=...` or env var BUILD_VERBOSE
GEN build/genhdr/mpversion.h
QSTR updated
GEN build/genhdr/qstrdefs.generated.h
Module registrations updated
GEN build/genhdr/moduledefs.h
Root pointer registrations updated
GEN build/genhdr/root_pointers.h
Traceback (most recent call last):
  File "/home/aappleby/circuitpython/mpy-cross/../py/maketranslationdata.py", line 29, in <module>
    import huffman
ModuleNotFoundError: No module named 'huffman'
make: *** [../py/py.mk:273: build/genhdr/compressed_translations.generated.h] Error 1
make: Leaving directory '/home/aappleby/circuitpython/mpy-cross'

**Reply by tannewt** [reply]
Our best build instructions are here: https://learn.adafruit.com/building-circuitpython

You can get help in #circuitpython-dev on Discord: https://adafru.it/discord I'm there during the work week.

Our build isn't simple so it may not be worth your time...
